### PR TITLE
Refactoring and extending update all products test suit

### LINKED ISSUE #1457: Add more UpdateAllProduct tests.
Some of the tests we might want to add:

- Filter out some products from the batches and confirm that batches continue even if some batch sizes are smaller or empty, see #861
- Test a final batch smaller than batch size (this will identify that `handle_complete` is not always called, which isn't an issue because we never use it)

_Originally posted by @mikkamp in https://github.com/woocommerce/google-listings-and-ads/pull/1451#pullrequestreview-955934017_

### PR DESCRIPTION
**Pull request by message-dimke**
### Changes proposed in this Pull Request:

Refactoring and extending update all products tests suit by adding tests to check multiple half filled or empty batches does continue batch scheduling and items processing as expected.

Closes #1457.

### Changelog entry

> Add - Extending Update All Products Test Suit

### CODE REVIEW

**Comment by message-dimke** [comment]
Also I found that when cherrypicking hotfix I did not pick my last updates to the code and styling so those appear into this PR as well.

File: tests/Unit/Jobs/UpdateAllProductsTest.php [hunk 1]
```diff
@@ -64,68 +61,216 @@ public function setUp(): void {
...
+			->willReturn( $filtered_product_list );
 
-		$this->filtered_product_list->expects( $this->never() )
-			->method( 'get' );
+		/*
+		 * We expect only single call to `has_scheduled_action` when scheduling
+		 * the batch and no calls further since batch is empty.
+		 */
+		$this->action_scheduler->expects( $this->once() )
+			->method( 'has_scheduled_action' );
```
Comment (mikkamp) [inline tests/Unit/Jobs/UpdateAllProductsTest.php:106]: I know the order isn't important, but when I was trying to mentally follow what code is executed and what we are expecting, it seemed reversed.

When we first call `job->schedule` it will run:
- `has_scheduled_action`
- `schedule_immediate`

When we call the `do_action` for the batch hook then it will run:
- `find_sync_ready_products`
- `has_scheduled_action`
- `schedule_immediate`

So it would seem logical to put the expectation of `has_scheduled_action` before `schedule_immediate`.
(same comment for the test `test_single_partial_batch`)

File: tests/Unit/Jobs/UpdateAllProductsTest.php [hunk 2]
```diff
@@ -64,68 +61,216 @@ public function setUp(): void {
...
 		$this->action_scheduler->expects( $this->exactly( 2 ) )
+			->method( 'has_scheduled_action' );
+
+		$this->job->schedule();
+
+		do_action( self::CREATE_BATCH_HOOK, 1 );
+	}
+
+	/**
+	 * We test two filly loaded batches to make sure batch and process item jobs are correctly scheduled
```
Comment (mikkamp) [inline tests/Unit/Jobs/UpdateAllProductsTest.php]: Just a small typo `fully loaded`. Also when we refer to batch and process items, these are separate actions within the full job.
```suggestion
	 * We test two fully loaded batches to make sure batch and process item actions are correctly scheduled
```

File: tests/Unit/Jobs/UpdateAllProductsTest.php [hunk 3]
```diff
@@ -64,68 +61,216 @@ public function setUp(): void {
...
+
+		/* adding a filter to make batch smaller for testing */
+		add_filter(
+			'woocommerce_gla_batched_job_size',
+			function ( $batch_count, $job_name ) {
+				if ( self::JOB_NAME === $job_name ) {
+					return 2;
+				}
+				return $batch_count;
+			}, 10, 2 );
```
Comment (mikkamp) [inline tests/Unit/Jobs/UpdateAllProductsTest.php]: If we split the parameters to multiple lines then the PHPCS rules we use require us to put each parameter on a separate line.
```suggestion
			},
			10,
			2
		);
```

**Review by mikkamp — APPROVED** [review]
Thanks for adding these additional tests. The tests look good and I'm getting the results as expected.

I just added a few small minor comments, but none of them are blockers, so I'll approve the PR.

File: tests/Unit/Jobs/UpdateAllProductsTest.php (hunk 1 above)
Comment (message-dimke) [reply tests/Unit/Jobs/UpdateAllProductsTest.php:106]: I changed the order of calls to have expectations aligned to how the corresponding methods are actually called.

File: tests/Unit/Jobs/UpdateAllProductsTest.php (hunk 2 above)
Comment (message-dimke) [reply tests/Unit/Jobs/UpdateAllProductsTest.php]: Fixed the typo and changed jobs to actions. Thanks!
